service/worker/esanalyzer/workflow.go

```diff
@@ -0,0 +1,410 @@
+		}
+	}
+
+	return nil
+}
+
+func getFindStuckWorkflowsQuery(
+	startDateTime int64,
+	endTime int64,
+	workflowType string,
+	maxNumWorkflows int,
+) (string, error) {
+	wfTypeMarshaled, err := json.Marshal(workflowType)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf(`
+    {
+      "query": {
+          "bool": {
+              "must": [
+                  {
+                      "range" : {
+                          "StartTime" : {
+                              "gte" : "%d",
+                              "lte" : "%d"
+                          }
+                      }
+                  },
+                  {
+                      "match" : {
+                          "WorkflowType" : "%s"
+                      }
+                  }
+              ],
+              "must_not": {
+                "exists": {
+                  "field": "CloseTime"
+                }
+              }
+          }
+      },
+      "size": %d
+    }
+    `, startDateTime, endTime, wfTypeMarshaled, maxNumWorkflows), nil
+}
+
+// findStuckWorkflows is activity to find open workflows that are live significantly longer than average
+func findStuckWorkflows(ctx context.Context, info WorkflowTypeInfo) ([]WorkflowInfo, error) {
+	analyzer := ctx.Value(analyzerContextKey).(*Analyzer)
+	logger := activity.GetLogger(ctx)
+
+	minNumWorkflowsNeeded := int64(analyzer.config.ESAnalyzerMinNumWorkflowsForAvg(info.Name))
+	if len(analyzer.config.ESAnalyzerLimitToTypes()) > 0 {
+		logger.Info("Skipping minimum workflow count validation since workflow types were passed from config")
+	} else if info.NumWorkflows < minNumWorkflowsNeeded {
+		logger.Warn(fmt.Sprintf(
+			"Skipping workflow type '%s' because it doesn't have enough(%d) workflows to avg",
+			info.Name,
+			minNumWorkflowsNeeded,
+		))
+		return nil, nil
+	}
+
+	startDateTime := time.Now().Add(-analyzer.config.ESAnalyzerTimeWindow()).UnixNano()
+
+	// allow some buffer time to any workflow
+	maxEndTimeAllowed := time.Now().Add(-analyzer.config.ESAnalyzerBufferWaitTime(info.Name)).UnixNano()
+
+	// if the workflow exec time takes longer than 3x avg time, we refresh
+	endTime := time.Now().Add(
+		-time.Nanosecond * time.Duration((int64(info.Duration.AvgExecTimeNanoseconds) * 3)),
+	).UnixNano()
+	if endTime > maxEndTimeAllowed {
+		endTime = maxEndTimeAllowed
+	}
+
+	maxNumWorkflows := analyzer.config.ESAnalyzerNumWorkflowsToRefresh(info.Name)
+	query, err := getFindStuckWorkflowsQuery(startDateTime, endTime, info.Name, maxNumWorkflows)
+	if err != nil {
+		logger.Error("Failed to create ElasticSearch query for stuck workflows",
+			zap.String("error", fmt.Sprintf("%v", err)),
+			zap.String("startDateTime", fmt.Sprintf("%v", startDateTime)),
+			zap.String("endTime", fmt.Sprintf("%v", endTime)),
+			zap.String("workflowType", info.Name),
+			zap.String("maxNumWorkflows", fmt.Sprintf("%v", maxNumWorkflows)),
```

Comment:
```suggestion
			zap.Int64("startDateTime", startDateTime),
			zap.Int64("endTime", endTime),
			zap.String("workflowType", info.Name),
			zap.Int64("maxNumWorkflows", maxNumWorkflows),
```
at least, I think that func's available.

though more generally this is why I'd generally keep things in durations, and `.UnixNano()` at the call-site.  that way you can log the `.String()` or format it or use `.Unix()` since nanoseconds probably don't matter much, or whatever.